### ISSUE TITLE
List indentation should not depend on marker width

### ISSUE THREAD

**Issue by Gravitonic** [typst/typst#4126]
### Description

Because the layout of lists (`list` and `enum`) are adjusted by the spacing before and after markers, the width of the marker itself can affect the indentation of the list. Since different lists' markers can have different widths, this means that lists can have different indentation:

<details><summary>Screenshot</summary>
<p>

<img width="430" alt="Untitled" src="https://github.com/typst/typst/assets/29341383/3d533983-68d4-4864-a9fd-2c698222eb50">

</p>
</details>

<details><summary>Typst code to reproduce</summary>
<p>

```typst
+ one
+ two
+ three

intervening paragraph

11. eleven
+ twelve
+ thirteen
```

</p>
</details> 

In contrast, LaTeX, Google Docs, and Microsoft Word all seem to let users set the indentation based on the list item bodies.

<details><summary>LaTeX</summary>
<p>

<img width="278" alt="Screenshot 2024-05-13 at 9 32 14 AM" src="https://github.com/typst/typst/assets/29341383/4156ae77-148d-4638-852e-b6cfd027b7c6">

</p>
</details> 

<details><summary>Google Docs</summary>
<p>

![image](https://github.com/typst/typst/assets/29341383/a08c326d-f58e-48f9-9977-5b2cf1cfc4b3)

</p>
</details> 

<details><summary>Microsoft Word</summary>
<p>

![image](https://github.com/typst/typst/assets/29341383/774b10d1-4e2b-433b-87c1-5bb30b9481c8)

</p>
</details> 

### Reproduction URL

https://typst.app/project/rqnPOljOs8eDzgh36DKpki

### Operating system

Web app, macOS

### Typst version

- [X] I am using the latest version of Typst

**Comment by Enter-tainer** [comment]
In LaTeX, the list markers are right aligned, and there is a minimal gap between list marker and actual content.

**Comment by Enter-tainer** [comment]
When list marker is left aligned, the dot will not be aligned:

```
\usepackage{enumitem}
\newcommand\setItemnumber[1]{\setcounter{enumi}{\numexpr#1-1\relax}}

\begin{enumerate}[align=left]
    \item first
    \item second
    \setItemnumber{500000000000}
    \item very large
\end{enumerate}
```

![image](https://github.com/typst/typst/assets/25521218/23084c9a-c303-4de0-be98-f065fb3c3c1e)

**Comment by Enter-tainer** [comment]
Microsoft word:

![image](https://github.com/typst/typst/assets/25521218/af72a623-0d6f-42f5-b0b3-7e2a6021be01)

**Comment by Enter-tainer** [comment]
Overall I feel like both word, typst and LaTeX's solution is good. Currently typst can align list content and list marker using grids, at the cost of baseline mismatch. It would be better if list marker in typst can be in the paragraph as list content, so they share the same baseline naturally.

For me i think normal people will never notice the alignment difference: they will never have a super long list that has more than 100 items.

**Comment by Enter-tainer** [comment]
For this case, I think we can first measure the maximum width of all list markers. Then we can correctly set the width of the box. So we can beat LaTeX and word 😎

**Comment by Enivex** [comment]
> For this case, I think we can first measure the maximum width of all list markers. Then we can correctly set the width of the box. So we can beat LaTeX and word 😎 

I think that tex and word is choosing a middle ground on purpose. Imagine you had one label that's much longer than the other ones.

**Comment by Enter-tainer** [comment]
> > For this case, I think we can first measure the maximum width of all list markers. Then we can correctly set the width of the box. So we can beat LaTeX and word 😎
> 
> I think that tex and word is choosing a middle ground on purpose. Imagine you had one label that's much longer than the other ones.

It looks like the default in tex and word can fit in 3~4 numbers. I think this is a good default and this covers most use case. The width can also be customized.

**Comment by Enter-tainer** [comment]
Comparison between LaTeX and word for lists with longer content. @laurmaedje what do you think of this? Will it make sense if we switch enum/list implementation to something similar to this?

![image](https://github.com/typst/typst/assets/25521218/65e27fef-8ba2-463c-8653-b8f0831e3064)
![image](https://github.com/typst/typst/assets/25521218/8acfbc8a-beb3-4a89-b471-e250ccf3eae7)

**Comment by laurmaedje** [comment]
Yes, I think we should switch lists to something like this. I believe Word uses tab stops to implement the behaviour. Without them or a min-width property, I think it will be hard to get this behaviour. I started a branch for tab stops recently.

**Comment by Enter-tainer** [comment]
>  I started a branch for tab stops recently.

Sounds cool! I'd like to do some initial work by reimplementing list/enum in a term-list-like way. And I guess I can somehow use your tab stop feature when it is merged.
